### PR TITLE
create Servers object

### PR DESCRIPTION
**Pull request by rmelian**
I'm creating this PR to start getting feedback.

This includes the Server object but still missing the Security object.

### CODE REVIEW

File: models/servers.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	if value.Extensions == nil {
+		return json.Marshal(value)
+	}
+	return MarshalWithExtensions(value, value.Extensions)
+}
+
+// ServerVariables object
+type ServerVariables struct {
+	AdditionalProperties map[string]*ServerVariable `json:",omitempty"`
+}
```
Comment (fmvilas) [inline models/servers.go]: You can avoid having this weird "AdditionalProperties" by doing the following:

```go
type ServerVariables map[string]*ServerVariable
```

Then you reference it above:

```go
Variables     *ServerVariables `json:"variables,omitempty"`
```

See this example: https://github.com/asyncapi/parser/blob/master/models/extension.go#L10

File: models/servers.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	Scheme        string           `json:"scheme"`
+	SchemeVersion string           `json:"schemeVersion,omitempty"`
+	Variables     *ServerVariables `json:"variables,omitempty"`
+	BaseChannel   string           `json:"baseChannel,omitempty"`
+	// Security      []*SecurityRequirement `json:"security,omitempty"`
+}
+
+// Unmarshal unmarshals JSON
+func (value *Server) Unmarshal(data []byte) error {
+	return json.Unmarshal(data, value)
```
Comment (fmvilas) [inline models/servers.go]: We should avoid calling `json.Unmarshal` inside the Unmarshal method. I know I've done it in the Info object but now we know there's the other way of doing it. This method should just create and return a new map and add property values one by one. If we do it and also rename the method to `UnmarshalJSON` the `json.Unmarshal` function will automatically call it.

File: models/servers.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// Marshal marshals JSON
+func (value *Server) Marshal() ([]byte, error) {
+	if value.Extensions == nil {
+		return json.Marshal(value)
+	}
+	return MarshalWithExtensions(value, value.Extensions)
+}
```
Comment (fmvilas) [inline models/servers.go]: We should avoid calling `json.Marshal(value)` inside the Marshal method. I know I've done it in the Info object but now we know there's the other way of doing it. This method should just create a map, call `json.Marshal(theNewMap)`, and return the resulting byte array. If we do it and also rename the method to `MarshalJSON` the `json.Marshal` function will automatically call it.

File: models/servers.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		if err := json.Unmarshal([]byte(v), &additionalValue); err != nil {
+			return err // invalid additionalProperty
+		}
+		if strct.AdditionalProperties == nil {
+			strct.AdditionalProperties = make(map[string]*ServerVariable, 0)
+		}
+		strct.AdditionalProperties[k] = additionalValue
+	}
+	return nil
+}
```
Comment (fmvilas) [inline models/servers.go]: Ok, so this is exactly what I suggested above 😄 👏

File: models/servers.go (hunk 1 above)
Comment (rmelian) [reply models/servers.go]: great, I knew there had to be a better way :)

File: models/servers.go (hunk 2 above)
Comment (rmelian) [reply models/servers.go]: great I wanted to get your confirmation on this, 100% agree

**Comment by fmvilas** [comment]
I think we can close this PR now, right?

**Pull request closed without merging**